File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/AgentThrottle.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;
import java.util.UUID;

/**
 * AgentThrottle
 *
 * <p>Template: {@code AgentThrottle Low 81 NotTrusted Zerocoded}
 * (recovered/reference/message_template.msg).
 */
public class AgentThrottle extends SLMessage {
    public AgentData AgentData_Field;
    public Throttle Throttle_Field;

    /** Block AgentData, Single. */
    public static class AgentData {
        public UUID AgentID; // LLUUID
        public int CircuitCode; // U32
        public UUID SessionID; // LLUUID
    }

    /** Block Throttle, Single. */
    public static class Throttle {
        public int GenCounter; // U32
        public byte[] Throttles; // Variable 1
    }

    public AgentThrottle() {
        this.zeroCoded = true;
        this.AgentData_Field = new AgentData();
        this.Throttle_Field = new Throttle();
    }

    @Override
    public int CalcPayloadSize() {
        return this.Throttle_Field.Throttles.length + 5 + 40;
    }

    @Override
    public void Handle(SLMessageHandler sLMessageHandler) {
        sLMessageHandler.HandleAgentThrottle(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 81 (AgentThrottle).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x00);
        byteBuffer.put((byte) 0x51);
        packUUID(byteBuffer, this.AgentData_Field.AgentID);
        packUUID(byteBuffer, this.AgentData_Field.SessionID);
        packInt(byteBuffer, this.AgentData_Field.CircuitCode);
        packInt(byteBuffer, this.Throttle_Field.GenCounter);
        packVariable(byteBuffer, this.Throttle_Field.Throttles, 1);
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.AgentData_Field.AgentID = unpackUUID(byteBuffer);
        this.AgentData_Field.SessionID = unpackUUID(byteBuffer);
        this.AgentData_Field.CircuitCode = unpackInt(byteBuffer);
        this.Throttle_Field.GenCounter = unpackInt(byteBuffer);
        this.Throttle_Field.Throttles = unpackVariable(byteBuffer, 1);
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/AgentThrottle.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.nio.ByteBuffer
import java.util.UUID

/**
 * AgentThrottle
 *
 * <p>Template: {@code AgentThrottle Low 81 NotTrusted Zerocoded}
 * (recovered/reference/message_template.msg).
 */
open class AgentThrottle : SLMessage() {
    @JvmField var AgentData_Field: AgentData = AgentData()
    @JvmField var Throttle_Field: Throttle = Throttle()

    /** Block AgentData, Single. */
    open class AgentData {
        @JvmField var AgentID: UUID? = null
        @JvmField var CircuitCode: Int = 0
        @JvmField var SessionID: UUID? = null
    }

    /** Block Throttle, Single. */
    open class Throttle {
        @JvmField var GenCounter: Int = 0
        @JvmField var Throttles: ByteArray? = null
    }

    init {
        zeroCoded = true
    }

    override fun CalcPayloadSize(): Int {
        return Throttle_Field.Throttles!!.size + 5 + 40
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleAgentThrottle(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 81 (AgentThrottle).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x00).toByte())
        byteBuffer.put((0x51).toByte())
        packUUID(byteBuffer, AgentData_Field.AgentID)
        packUUID(byteBuffer, AgentData_Field.SessionID)
        packInt(byteBuffer, AgentData_Field.CircuitCode)
        packInt(byteBuffer, Throttle_Field.GenCounter)
        packVariable(byteBuffer, Throttle_Field.Throttles, 1)
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        AgentData_Field.AgentID = unpackUUID(byteBuffer)
        AgentData_Field.SessionID = unpackUUID(byteBuffer)
        AgentData_Field.CircuitCode = unpackInt(byteBuffer)
        Throttle_Field.GenCounter = unpackInt(byteBuffer)
        Throttle_Field.Throttles = unpackVariable(byteBuffer, 1)
    }
}
